src/Stratis.Bitcoin/Consensus/ConsensusManager.cs

```diff
@@ -414,67 +414,89 @@ private async Task OnPartialValidationSucceededAsync(ChainedHeader chainedHeader
         {
             this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeader), chainedHeader);
 
-            List<ChainedHeaderBlock> chainedHeaderBlocksToValidate;
-            ConnectBlocksResult connectBlocksResult = null;
+            bool shouldPartialValidationContinue = false;
+            ConnectBlocksResult fullResult = null;
+            PartialValidationSucceededResult partialResult = null;
 
             using (await this.reorgLock.LockAsync().ConfigureAwait(false))
             {
-                bool fullValidationRequired;
-
                 lock (this.peerLock)
                 {
-                    chainedHeaderBlocksToValidate = this.chainedHeaderTree.PartialValidationSucceeded(chainedHeader, out fullValidationRequired);
+                    partialResult = this.chainedHeaderTree.PartialValidationSucceeded(chainedHeader);
+                    if (partialResult.ChainedHeaderBlocksToValidate == null)
```

Comment:
if `partialResult.ChainedHeaderBlocksToValidate == null` we may still need to fully validate so returning here seems wrong

Reply:
it seems wrong (though i believe it is not, just seemingly), that's why i think the new code is actually worse. if you spend 5 minutes on it (which you should not), you will see that it is probably correct as this has a special meaning. It's comment says 

```
/// <summary>List of chained header blocks with block data that should be partially validated next. Or <c>null</c> if none should be validated.</summary>
``` 

but it is hardly obvious from that that not even the current block should not be validated (that it implies that full validation is not needed.

Reply:
But this is a classic case of the tip, a new block arrived it needs to be fully validated but no more blocks after that require partial validation (and this is similar to the way it worked before)

Reply:
i know, but the new logic is that if CHBTV is null it means no FV, but if it is not null, just empty list, then you may do FV. you don't need to convince me that it smells though :)

Reply:
@Aprogiena ah ok now it makes sense. Let me fix this.